### ISSUE TITLE
TLS troubleshooting and migration guide

### ISSUE THREAD

**Issue by indirect** [bundler/bundler#5687]
As part of [the Fastly TLS deprecation][1], RubyGems.org is going to lose the ability to allow TLS v1.0 and v1.1. A few weeks ago, Fastly disabled TLS v1.0 and v1.1 for us as part of their plan, and we discovered that _a lot_ of Ruby users still aren't able to use TLS v1.2 yet. Unfortunately, the older versions of TLS have known problems that make them insecure.

We would like to migrate everyone, and Fastly is going to force all of their customers to migrate no later than April 2018. We [plan to disable older TLS versions in January 2018][2], with the goal of releasing a new version with a warning that explains how to fix the problem it at least 3 months in advance.

In order to be able to use TLS v1.2, users need to be using OpenSSL version 1.0.1 or later (first released March 14, 2012). Then, they need to make sure that they are using a version of Ruby that includes OpenSSL support, and binds against OpenSSL 1.0.1 or later. Unfortunately, there are a lot of ways that can be messed up, and problems with OpenSSL versions can interact with problems with certificates, sometimes making it hard to tell exactly where the problem lies.

I am currently imagining that our explanation of the problem and possible solutions will take the form of a guide on [the Bundler website](https://bundler.io). I'm hoping we can create a page that explains what the problem is, why we are changing things, and how to update the OpenSSL library and/or Ruby to fix the problem. Then, we can link to that page from the warning that we print when Bundler is used on a version of Ruby that will not be able to connect to RubyGems.org after the change.

Separate from the migration to TLS v1.2, but related in that it causes a similar problem is general SSL and certificate issues. On the whole, even without the TLS version change, these have been the source of a lot of trouble. rubygems/rubygems#1758 is a great example of the kinds of problems that people have with SSL in general, and the kinds of things that they need to do to get it working again. We have [a RubyGems guide to solving SSL problems][3], but it seems to be insufficient based on how many people still open new issues or comment on old ones like the ticket I linked above. The source for the RubyGems SSL guide is in the rubygems/guides repo [in this file][4].

Another super helpful resource that I refer to a lot while troubleshooting SSL problems with Ruby is [Mislav's guide to Ruby and SSL](https://mislav.net/2013/07/ruby-openssl/). It's a good explanation of the kinds of problems that Ruby devs often see with SSL (although it doesn't mention the TLS version migration, because that is a newer problem than the guide). He also wrote [a Ruby to help debug SSL issues](https://github.com/mislav/ssl-tools), which is another helpful resource.

In the Bundler guide, we'll probably need to not only explain the TLS v1.2 migration, but also include most of the information from the RubyGems guide (and even additional information above and beyond that guide, like what users reported in that ticket). I'm sure users having any kind of SSL problem will end up checking out our guide, and if we can successfully help them solve their own problems, that will be a huge win.

I'm sure you'll have questions once you start tackling this, feel free to ask here or ping me in Slack for more details. Thanks!

[1]: https://www.fastly.com/blog/phase-two-our-tls-10-and-11-deprecation-plan
[2]: https://github.com/bundler/bundler/pull/5638
[3]: http://guides.rubygems.org/ssl-certificate-update/
[4]: https://github.com/rubygems/guides/blob/gh-pages/ssl-certificate-update.md

**Comment by rubymorillo** [comment]
Hey!

So I've started reading all of the supporting material and I think I'm ready to get started. Here are a few q's:

1. It sounds like we want to incorporate this guide into the new Bundler guide: [SSL Cert Update](https://github.com/rubygems/guides/blob/gh-pages/ssl-certificate-update.md)
The steps outlined in that guide are as follows:

Step 1: Obtain new trust certificate
Step 2: Locate RubyGems cert directory in installation
Step 3: Copy new trust certificate

2. Is this guide^^^ a prerequisite to the Bundler migration guide (as in, should we link out to it), or are we trying to incorporate the above guide into this new, Bundler migration guide?

3. You mention how “In order to use TLS v1.2, users need to be using OpenSSL version 1.0.1 or later and to ensure they are using a version of Ruby that includes OpenSSL support. This sounds like the first step in the process. Can you outline what this actually looks like in terms of actions users need to take to do this? (bullet list is fine: i.e. Step 1, use this command in your CLI to check OpenSSL version, step 2, if you’re using v1.x go to [website] to download new version, etc etc). I’ll in turn flesh these out in the guide.

4. What are the Bundler-specific steps and commands we want for the migration?

I think once I have the above information, I can start taking a stab on a first draft for you and the team to review, provided other q’s don’t come up when I get your responses!

**Comment by rubymorillo** [comment]
I thought of a few more @indirect:

5. How much explaining do we have to do of how TLS certs work with gems beforehand? (I'm guessing the guide itself won't be the best place for this, but is this a source of confusion in and of itself?)

6. Are there FAQ that this guide *must* answer? I know we want it to touch the points you've raised in the issue, but are there specific questions that might be easily missed that we didn't include here? Thx again!

**Comment by indirect** [comment]
@rubymorillo here you go!

> Is [the SSL Cert Update Guide] a prerequisite to the Bundler migration guide (as in, should we link out to it), or are we trying to incorporate the above guide into this new, Bundler migration guide?

It is not a prerequisite. I expect that users who are having trouble with an old OpenSSL version are also likely to have trouble with their SSL certs, and therefore need that guide as well. As a result, I think we should either refer them to that guide or preferably update that guide and include it (as an optional aside) in the new TLS migration guide.

In an ideal world, the issues that the SSL Cert Guide tries to help with would always be solved by installing the latest Bundler or RubyGems. For complicated reasons, sometimes that isn’t enough, and we’ve basically written that guide by combining all of the troubleshooting steps that people report have helped them. There are several new suggestions that we would like to add to the SSL cert guide as we update it, mostly reported by people in this thread: [rubygems/rubygems#1758](https://github.com/rubygems/rubygems/issues/1758 "rubygems/rubygems#1758")

> You mention how “In order to use TLS v1.2, users need to be using OpenSSL version 1.0.1 or later and to ensure they are using a version of Ruby that includes OpenSSL support. This sounds like the first step in the process. Can you outline what this actually looks like in terms of actions users need to take to do this?

Unfortunately, it varies wildly depending on not just the exact OS that a user is running, but also how they have installed both Ruby and OpenSSL. :( For example, I can give you steps for someone on a Mac, who installed OpenSSL with Homebrew, and then installed Ruby with ruby-install. There are completely separate steps for someone on a Mac who used RVM to install Ruby. We’re probably going to need to create a huge list of the many ways to update your OpenSSL and then recompile Ruby, and I don’t know the specific steps needed for most of the possible combinations.

Off the top of my head, I expect people will ask us for instructions for each combination of:

1. operating system, which could be: macOS, Windows, Debian/Ubuntu, Fedora/RedHat/CentOS, Arch, Gentoo, maybe even FreeBSD
2. ruby installation method, which could be: OS package manager, rvm, ruby-build, ruby-install, pik, RubyInstaller2, compiled from source
3. openssl installation method, which could be: OS package manager, rvm, ruby-build, compiled from source

As you can see, that’s… a lot of possibilities. Maybe more than a hundred. 😬 I don’t think it’s realistic for us to write and test that many combinations, so here are the ones that I think we will need to be able to announce the migration:

1. macOS with ruby+openssl from rvm
2. macOS with ruby+openssl rbenv/ruby-build
3. macOS with ruby from chruby/ruby-install and openssl from homebrew
4. Debian/Ubuntu with ruby+openssl from rvm
5. Debian/Ubuntu with ruby+openssl from rbenv/ruby-build
6. Fedora/RedHat/CentOS with ruby+openssl from rvm
7. Fedora/RedHat/CentOS with ruby+openssl from rbenv/ruby-build
8. Windows with ruby+openssl from RubyInstaller

I can give individual steps for some of those, but we’ll probably need to make it a team effort across the entire Bundler+RubyGems teams to figure out the steps for all of those options.

> What are the Bundler-specific steps and commands we want for the migration?

Sadly, there aren’t any. Bundler is entirely at the mercy of Ruby and OpenSSL about whether it will be able to connect to RubyGems.org after the cutoff date.

> How much explaining do we have to do of how TLS certs work with gems beforehand? (I'm guessing the guide itself won't be the best place for this, but is this a source of confusion in and of itself?)

It seems like this is a source of confusion, yes, and I would love to include a section or aside or optional secondary guide that explains this.

> Are there FAQ that this guide must answer? I know we want it to touch the points you've raised in the issue, but are there specific questions that might be easily missed that we didn't include here? Thx again!

My imagined FAQ goes like this:

**What’s happening?**
RubyGems.org will be requiring TLS v1.2 or higher in January 2018.

**What does that mean?**
It means that installing gems using `gem install` or `bundle install` will require that your Ruby installation be able to use TLS v1.2. The TLS v1.2 protocol is provided by the OpenSSL gem, which uses the OpenSSL library written in C.

**Why are you doing this so abruptly?**
We don’t think it’s abrupt. OpenSSL has supported TLS v1.2 since March 2012. Versions 1.0 and 1.1 have been shown repeatedly to be dangerously insecure, and it’s not safe to use them anymore.

**This means I can’t install gems!**
If you are using a Ruby linked against OpenSSL from before March 2012, we’re providing a community-maintained guide to help you upgrade. If the guide doesn’t work for you, please open a ticket and we’ll try to help you.

**Your guide doesn’t work for me!**
That’s not a question, but we would be happy to help you if you open a new ticket at \<some repo\>, and provide the following information:

Operating system and version:
Ruby version:
RubyGems version:
Bundler version:
Output from running `ruby -ropenssl -e 'puts OpenSSL::OPENSSL_LIBRARY_VERSION'`: 
Ruby version manager (if any):
OpenSSL installed from (if any):

**Comment by rubymorillo** [comment]
Hey hey @indirect!

Thanks for this!! It gives me a lot of context. For next steps on my end, here's what I'll do:

- Craft an intro based on the imagined FAQ above ^^^
- Include a link to the SSL cert guide in the TLS migration guides (Separately, it sounds like the SSL guide should get updated once we create this one, but if you think it's the other way around, let me know.)

What we'll need is:
- *From the team*. Figuring out who will be the right people to add steps to each of the following, and get them to send us this information (perhaps as a comment in this issue):
1. macOS with ruby+openssl from rvm
2. macOS with ruby+openssl rbenv/ruby-build
3. macOS with ruby from chruby/ruby-install and openssl from homebrew
4. Debian/Ubuntu with ruby+openssl from rvm
5. Debian/Ubuntu with ruby+openssl from rbenv/ruby-build
6. Fedora/RedHat/CentOS with ruby+openssl from rvm
7. Fedora/RedHat/CentOS with ruby+openssl from rbenv/ruby-build
8. Windows with ruby+openssl from RubyInstaller
- *From you*. More info about this:
1. How much explaining do we have to do of how TLS certs work with gems beforehand?

Let me know if there's anything I'm missing. Feel free to send me whatever information you can and I'll fill those sections in in the meanwhile 😄. I'll work on an outline and will send to you either tomorrow or Wednesday when I'm done.

**Comment by rubymorillo** [comment]
@indirect here's the outline:

**Introduction**
- 2-3 sentences about Fastly deprecating older versions and what this means for RubyGems 
   - Date when RubyGems.org will move over to TLSv1.2 and up
   - What this looks like
- What prerequisites they should have (i.e. Ruby, rvm, rbenv, OpenSSL versions, etc)
- Mention the setups covered in this guide
- A separation section (like a separate, blue box) that gives some background into how TLS certs work with gems. 

**Instructions by setup** (we should also specify OS versions, if possible)

1. - macOS with ruby+openssl from rvm
2. - macOS with ruby+openssl rbenv/ruby-build
3. - macOS with ruby from chruby/ruby-install and openssl from homebrew
4. - Debian/Ubuntu with ruby+openssl from rvm
5. - Debian/Ubuntu with ruby+openssl from rbenv/ruby-build
6. - Fedora/RedHat/CentOS with ruby+openssl from rvm
7. - Fedora/RedHat/CentOS with ruby+openssl from rbenv/ruby-build
8. - Windows with ruby+openssl from RubyInstaller

**What if my setup isn’t covered here?**

- Direct folks to where they can open a ticket [repo here] and include the following info:
   - Ruby version:
   - RubyGems version:
   - Bundler version:
   - Output from running ruby -ropenssl -e 'puts OpenSSL::OPENSSL_LIBRARY_VERSION':
   - Ruby version manager (if any):
   - OpenSSL installed from (if any):

**Next steps (link out to additional resources they should look at)**
- The SSL guide, why they might want to update that as well, and link to the guide
- Anything else?

Let me know what your thoughts are and if you think this rough outline covers what we need!

**Comment by rubymorillo** [comment]
Hey @indirect I've an update!

So I've made inroads with the SSL guide updates. I'll send you the working doc for that separately. 

Here's what I've done:

* **Separated out the actual tutorial/guide sections from the background sections**. I propose we turn the "background" and "specific problems with RubyGems" sections into an article which we can reference in the guide. Since most folks just want to know how to solve the problem, leaving those sections might add more friction. I've formatted it that way in the Google doc.
* **Cleaned up the language to make it easier to read.**
* **Added comments in the areas that we need more clarification.** I've also added some sentences and want to get someone to look at them for technical accuracy.
* **Added a few of the community-sourced solutions [we saw here](https://github.com/rubygems/rubygems/issues/1758)**. It would be great to get your eyes and @duckinator 's to ensure that we have the correct solutions in there. A few in the issue were referenced directly in other comments or had emojis which I took as a vote of confidence, but a few others didn't appear to have been tested out by anyone. So having a set of eyes on that would be great!

In short, next steps:
* Get your eyes on the doc to ensure language looks good, the added content is technically accurate, and that we're featuring the community-sourced solutions that were tested out by other users.

I'll send you the doc in short order. 

Thanks!

**Comment by indirect** [comment]
@rubymorillo That outline for the TLS guide looks great! After seeing the list of combinations written down, I think we'll need to adjust it:

1. macOS built-in ruby (solve by upgrading macOS to 10.13 High Sierra, not yet released)
2. macOS with ruby from chruby/ruby-install and openssl from homebrew (solve by upgrading openssl with homebrew)
4. any OS with ruby+openssl from rvm (solve with `rvm implode` and reinstall rvm)
5. any OS with ruby+openssl rbenv/ruby-build (solve by removing and reinstalling rubies)
7. debian/ubuntu with ruby installed from apt (I assume upgrade ruby and openssl with apt, but I haven't tested. I think @BanzaiMan can tell us how Travis did this.)
9. fedora/redhat/centos with ruby install from yum (I assume upgrade ruby and openssl with yum, but I don't actually know.)
11. Windows with ruby+openssl from RubyInstaller (solve by removing ruby and reinstalling with RubyInstaller)
12.

**Comment by duckinator** [comment]
Should I try to set up a server that only uses TLS 1.2+, and write a program that tries to connect to said server and tells you whether or not your version of Ruby + OpenSSL are compatible?

Also, you're doing fantastic work, @rubymorillo!

**Comment by colby-swandale** [comment]
@duckinator You don't nessesarily need to setup a server to check if Ruby + OpenSSL supports TLS 1.2. There are several ways to inspect the supported versions of TLS in the openssl library.

Ruby 1.9.3
```
irb(main):01:0> OpenSSL::SSL::SSLContext::METHODS
=> [:TLSv1, :TLSv1_server, :TLSv1_client, :SSLv3, :SSLv3_server, :SSLv3_client, :SSLv23, :SSLv23_server, :SSLv23_client]

irb(main):02:0> OpenSSL::SSL::SSLContext.new(:TLSv1_2)
ArgumentError: unknown SSL method `TLSv1_2'.
        from (irb):14:in `initialize'
        from (irb):14:in `new'
        from (irb):14
        from /usr/local/bin/irb:12:in `<main>'
```

Ruby 2.4.1
```
irb(main):003:0> OpenSSL::SSL::SSLContext::METHODS
=> [:TLSv1, :TLSv1_server, :TLSv1_client, :TLSv1_2, :TLSv1_2_server, :TLSv1_2_client, :TLSv1_1, :TLSv1_1_server, :TLSv1_1_client, :SSLv2, :SSLv2_server, :SSLv2_client, :SSLv3, :SSLv3_server, :SSLv3_client, :SSLv23, :SSLv23_server, :SSLv23_client]

irb(main):002:0> OpenSSL::SSL::SSLContext.new(:TLSv1_2)
=> #<OpenSSL::SSL::SSLContext:0x007f9754037f80 @cert=nil, @key=nil, @client_ca=nil, @ca_file=nil, @ca_path=nil, @timeout=nil, @verify_mode=nil, @verify_depth=nil, @renegotiation_cb=nil, @verify_callback=nil, @cert_store=nil, @extra_chain_cert=nil, @client_cert_cb=nil, @session_id_context=nil, @tmp_dh_callback=nil, @session_get_cb=nil, @session_new_cb=nil, @session_remove_cb=nil, @tmp_ecdh_callback=nil, @servername_cb=nil, @npn_protocols=nil, @alpn_protocols=nil, @alpn_select_cb=nil, @npn_select_cb=nil>
```

This will let write a simple script to check if TLS 1.2 supported in the ruby env.

**Comment by duckinator** [comment]
@colby-swandale thanks for the tip!

So, basically, this:

```ruby
#!/usr/bin/env ruby

require 'openssl'

def supports_tls_1_2?
  OpenSSL::SSL::SSLContext::METHODS.include?(:TLSv1_2)
end

puts "TLS v1.2 support: #{supports_tls_1_2?}"
```

---

I [uploaded it to a pastebin](http://sprunge.us/jiRd), so now you can do the Bad Thing™ and execute arbitrary code from the internet:

```
$ curl -s sprunge.us/jiRd | ruby  
TLS 1.2 support: true
$ wget -qO- sprunge.us/jiRd | ruby
TLS 1.2 support: true
$ 
```

**Comment by rubymorillo** [comment]
Yay @duckinator! Are we cool with me adding this in the guide somehow, prefacing that they can use this script to check whether or not a user's version of Ruby + OpenSSL is compatible? (And if so, is the pastebin link going to be permanent?)

**Comment by segiddins** [comment]
Maybe it would be better to suggest that users run `ruby -ropenssl -e 'puts "TLS v1.2 support: #{OpenSSL::SSL::SSLContext::METHODS.include?(:TLSv1_2)}"'` instead of the `curl | ruby` ?

**Comment by colby-swandale** [comment]
i agree with @segiddins

**Comment by rubymorillo** [comment]
Just so I'm clear about sequence, we'd want users to run this script^^^^ _after_ they've followed the guides to ensure their Ruby & OpenSSL versions are compatible, not before. Is that correct? @segiddins @colby-swandale @duckinator

**Comment by indirect** [comment]
@rubymorillo I think we would want to use it as a before-and-after check. If you're having problems, run that snippet to see if your problem is that you need to upgrade OpenSSL for TLSv1.2. If that code prints `true`, the problem is not that they need to upgrade to get TLSv1.2 support.

If that code prints `false`, the reader should follow the upgrade guide and then run the snippet again to verify that they have updated OpenSSL successfully.

**Comment by rubymorillo** [comment]
Hey @BanzaiMan, can you confirm the Debian/Ubuntu instructions for troubleshooting the TLS issue is to upgrade Ruby and OpenSSL w/`apt`? See comment here for more context: https://github.com/bundler/bundler/issues/5687#issuecomment-308252104

**Comment by rubymorillo** [comment]
@indirect perf, that's what I initially thought, too. Sounds good!

**Comment by duckinator** [comment]
@segiddins replaced the part of my comment that suggested doing `curl | ruby` with a link to your comment.

**Comment by colby-swandale** [comment]
@indirect can this be closed?
